### PR TITLE
Implement metadata image selection

### PR DESCRIPTION
**Pull request by Jetroid**
@dereklieu: Implements issue #998 without using an external service. #998 is an important issue for me and I cannot use prose without this feature.

Usage: 

```
prose:
  metadata:
    _posts:
      - name: "image"
        field:
          element: "image"
          placeholder: "default-image.jpg"
          label: "My Image"
```

Possible extension functionality:

Add a _prose.yml config value to allow configuration of it metadata is set equal to the filename, eg `background-homepage.png`, directory, eg `/assets/images/background-homepage.png` or fully qualified, eg `{{ site.baseurl }}/assets/images/background-homepage.png`.

### CODE REVIEW

**Comment by dereklieu** [comment]
Hey @Jetroid thanks for pulling in the time here. Just wanted to say that this is on my radar, even though I haven't looked at it yet, I plan to over the next day or two.

**Comment by ewiggin** [comment]
@dereklieu I'm waiting for this feature. 😄 What is the state of this?

**Comment by dereklieu** [comment]
My bad! Really blocking this but was buried under work for the past few months. I have some more time though and promise to look at this over the weekend.

File: app/util.js [hunk 1]
```diff
@@ -257,6 +257,42 @@ module.exports = {
     return false;
   },
 
+  renderMedia: function(data, elem, controller, back) {
```
Comment (dereklieu) [inline app/util.js]: I know this function is now repurposed to be called from two places, so I'm glad you put it in a util. However, I don't love the idea of passing around references to jQuery objects. Between different files, it gets a lot harder to know what's being referenced. Is there some way we can avoid this?

Comment (dereklieu) [inline app/util.js]: Since this is now in a different file, maybe `data` should be more descriptive.

File: app/views/file.js [hunk 2]
```diff
@@ -386,8 +386,9 @@ module.exports = Backbone.View.extend({
     // Bind Drag and Drop work on the editor
     if (this.model.get('markdown') && this.model.get('writable')) {
       upload.dragDrop(this.$el, (function(e, file, content) {
-        if (this.$el.find('#dialog').hasClass('dialog')) {
-          this.updateImageInsert(e, file, content);
+        var $dialog = this.$el.find('div[data-element="dialog"]');
+        if ($dialog.hasClass('dialog')) {
+          this.updateImageInsert(e, file, content, this.toolbar, dialog);
```
Comment (dereklieu) [inline app/views/file.js]: Seems like the argument should be `$dialog` and not `dialog`.

File: app/views/file.js [hunk 3]
```diff
@@ -1332,16 +1335,16 @@ module.exports = Backbone.View.extend({
...
   },
 
-  updateImageInsert: function(e, file, content) {
-    var path = (this.toolbar.mediaDirectoryPath) ?
-                    this.toolbar.mediaDirectoryPath :
-                    util.extractFilename(this.toolbar.file.attributes.path)[0];
-    var src = path + '/' + encodeURIComponent(file.name);
+  updateImageInsert: function(e, file, content, target, $dialog) {
+    var path = (this.config.media) ?
+                    this.config.media :
```
Comment (dereklieu) [inline app/views/file.js]: 👍

File: app/views/file.js [hunk 4]
```diff
@@ -1357,31 +1360,14 @@ module.exports = Backbone.View.extend({
...
-        this.editor.focus();
-        this.editor.replaceSelection(image + '\n', 'end');
-        this.updateSaveState('Saved', 'saved', true);
-      }).bind(this),
-      error: (function(model, xhr, options) {
-        var message = util.xhrErrorMessage(xhr);
-        this.updateSaveState(message, 'error');
-      }).bind(this)
+      success: (successCb).bind(this),
+      error: (failureCb).bind(this)
```
Comment (dereklieu) [inline app/views/file.js]: The `error` callbacks are all the same I think, so that can probably just stay put?

File: app/views/metadata.js [hunk 5]
```diff
@@ -380,6 +389,91 @@ module.exports = Backbone.View.extend({
...
+
+  updateInput: function(e, file, content) {
+    var $dialog = $(e.target, this.$el).closest(".dialog");
+    var $input = $dialog.parent().children("input");
+    var self = this;
+
+    if (this.queue) {
+      var userDefinedPath = $('input[name="url"]').val();
+
+      var onSuccess = function(model, res, options) {
```
Comment (dereklieu) [inline app/views/metadata.js]: I would guess that this is where it's failing to set the save state, and causing the spinner to endlessly spin.

**Comment by dereklieu** [comment]
@Jetroid first off, sorry for my tardiness here. You've done some great work in a codebase that isn't exactly easy to work with (and is downright nasty by today's standards) so 👍 for that.

I left some notes here. Overall I think this is on the right path but I think there are some changes I'd like to see before merging this. The main thing not mentioned here is unit tests. I wonder if breaking out some of this functionality that's now in the file, toolbar, and metadata views could be better served living in a separate, new image uploader view? I think that might help avoid some of the `$.closest` calls, which I'm not terribly excited about. This would also help improve testability, which would probably be a real chore the way it is right now, though that is certainly not your fault.

Happy to hear your thoughts here and help out however I can. Thanks again.

**Comment by kylekirkby** [comment]
Hi all,

Any chance on getting this feature merged ?

Best regards,
Kyle

**Comment by Jetroid** [comment]
Hi @dereklieu, sorry, I completely forgot about Prose and this pull request until Kyle posted his reminder.

I've taken another stab at this, and have broken out a lot of the functionality into a new view, AssetSelection, which renders the contents of the dialog when picking/uploading an image. 

Hopefully this is more in line with the sort of thing you want for Prose.

**Comment by kylekirkby** [comment]
LGTM - thanks @Jetroid !

**Comment by neosilky** [comment]
Would love this! :smile:

**Comment by Jetroid** [comment]
Hey @dereklieu, I'd like to avoid us forgetting about this again, if possible!

**Comment by Jetroid** [comment]
Ping @dereklieu

**Comment by Jetroid** [comment]
Ping @dereklieu

**Comment by Jetroid** [comment]
Ping @dereklieu

**Comment by buhrmi** [comment]
https://www.youtube.com/watch?v=oJDGcxAf9D8

**Comment by dy** [comment]
Ping @dereklieu

**Pull request merged**